src/partitioning.c

```diff
@@ -21,63 +34,109 @@ partitioning_func_set_func_fmgr(PartitioningFunc *pf)
 	if (funclist == NULL || funclist->next)
 		elog(ERROR, "Could not resolve the partitioning function");
 
+	pf->paramtype = funclist->args[0];
+
+	if (!(funclist->nargs == 1 &&
+		  (pf->paramtype == TEXTOID || pf->paramtype == ANYELEMENTOID)))
+		elog(ERROR, "Invalid partitioning function signature");
+
 	fmgr_info_cxt(funclist->oid, &pf->func_fmgr, CurrentMemoryContext);
 }
 
-static void
-partitioning_info_set_textfunc_fmgr(PartitioningInfo *pi, Oid relid)
+static Oid
+find_coercion_func(Oid type)
 {
-	Oid			type_id,
-				func_id;
-	bool		isVarlena;
+	Oid			funcid;
+	bool		is_varlena;
 	CoercionPathType cpt;
 
-	pi->column_attnum = get_attnum(relid, pi->column);
-	type_id = get_atttype(relid, pi->column_attnum);
-
 	/*
 	 * First look for an explicit cast type. Needed since the output of for
 	 * example character(20) not the same as character(20)::text
 	 */
-	cpt = find_coercion_pathway(TEXTOID, type_id, COERCION_EXPLICIT, &func_id);
+	cpt = find_coercion_pathway(TEXTOID, type, COERCION_EXPLICIT, &funcid);
+
 	if (cpt != COERCION_PATH_FUNC)
-	{
-		getTypeOutputInfo(type_id, &func_id, &isVarlena);
-	}
-	fmgr_info_cxt(func_id, &pi->partfunc.textfunc_fmgr, CurrentMemoryContext);
+		getTypeOutputInfo(type, &funcid, &is_varlena);
+
+	return funcid;
+}
+
+/*
+ * Resolve the function to cast a value to text.
+ */
+static void
+partitioning_info_set_textfunc_fmgr(PartitioningInfo *pi)
+{
+	Oid			funcid = find_coercion_func(pi->column_type);
+
+	fmgr_info_cxt(funcid, &pi->partfunc.textfunc_fmgr, CurrentMemoryContext);
 }
 
 PartitioningInfo *
-partitioning_info_create(int num_partitions,
-						 const char *schema,
+partitioning_info_create(const char *schema,
 						 const char *partfunc,
 						 const char *partcol,
 						 Oid relid)
 {
 	PartitioningInfo *pi;
+	Oid			varcollid,
+				funccollid = InvalidOid;
+	Var		   *var;
+	FuncExpr   *expr;
 
 	pi = palloc0(sizeof(PartitioningInfo));
 	strncpy(pi->partfunc.name, partfunc, NAMEDATALEN);
 	strncpy(pi->column, partcol, NAMEDATALEN);
+	pi->column_attnum = get_attnum(relid, pi->column);
+	pi->column_type = get_atttype(relid, pi->column_attnum);
 
 	if (schema != NULL)
 		strncpy(pi->partfunc.schema, schema, NAMEDATALEN);
 
 	partitioning_func_set_func_fmgr(&pi->partfunc);
-	partitioning_info_set_textfunc_fmgr(pi, relid);
+	partitioning_info_set_textfunc_fmgr(pi);
+
+	pi->partfunc.func_fmgr.fn_extra = lookup_type_cache(pi->column_type, 0);
+
+	/*
+	 * Prepare a function expression for this function. The hash function
+	 * needs this to be able to resolve the type of the value to be hashed.
+	 */
+	varcollid = get_typcollation(pi->column_type);
+
+	var = makeVar(1,
+				  pi->column_attnum,
+				  pi->column_type,
+				  -1,
+				  varcollid,
+				  0);
+
+	expr = makeFuncExpr(pi->partfunc.func_fmgr.fn_oid, INT4OID, list_make1(var),
```

Comment:
Given that we are adding a FuncExpr and that has a cost are we even sure this method of hashing is faster?

Reply:
This FuncExpr is only to pass on type information and is a one time cost per transaction while string conversion happens at least twice per tuple in the old implementation.

Reply:
I would still like to see some actual performance numbers for a change of this magnitude.